halo2_proofs/src/plonk/shuffle/prover.rs

```diff
@@ -0,0 +1,241 @@
+    ///   and S_compressed = \theta^{m-1} S_0 + theta^{m-2} S_1 + ... + \theta S_{m-2} + S_{m-1},
+    pub(in crate::plonk) fn compress<'a, C>(
+        &self,
+        pk: &ProvingKey<C>,
+        params: &Params<C>,
+        theta: ChallengeTheta<C>,
+        advice_values: &'a [Polynomial<C::Scalar, LagrangeCoeff>],
+        fixed_values: &'a [Polynomial<C::Scalar, LagrangeCoeff>],
+        instance_values: &'a [Polynomial<C::Scalar, LagrangeCoeff>],
+    ) -> Result<Compressed<C>, Error>
+    where
+        C: CurveAffine<ScalarExt = F>,
+        C::Curve: Mul<F, Output = C::Curve> + MulAssign<F>,
+    {
+        // Closure to get values of expressions and compress them
+        let compress_expressions = |expressions: &[Expression<C::Scalar>]| {
+            pk.vk.domain.lagrange_from_vec(evaluate_with_theta(
+                expressions,
+                params.n as usize,
+                1,
+                fixed_values,
+                advice_values,
+                instance_values,
+                *theta,
+            ))
+        };
+
+        let (input_expressions_group, shuffle_expressions_group) = self
+            .0
+            .iter()
+            .map(|v| {
+                // Get values of input expressions involved in the shuffle and compress them
+                let input_expression = compress_expressions(&v.input_expressions);
+
+                // Get values of table expressions involved in the shuffle and compress them
+                let shuffle_expression = compress_expressions(&v.shuffle_expressions);
+                (input_expression, shuffle_expression)
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+            .unzip();
+
+        Ok(Compressed {
+            input_expressions_group,
+            shuffle_expressions_group,
+        })
+    }
+}
+
+impl<C: CurveAffine> Compressed<C> {
+    /// Given a Shuffle with input expressions, table expressions this method constructs the
+    /// grand product polynomial over the shuffle. The grand product polynomial
+    /// is used to populate the Product<C> struct. The Product<C> struct is
+    /// added to the Shuffle and finally returned by the method.
+    pub(in crate::plonk) fn commit_product(
+        self,
+        pk: &ProvingKey<C>,
+        params: &Params<C>,
+        beta: ChallengeBeta<C>,
+        gamma: ChallengeGamma<C>,
+        delta: ChallengeDelta<C>,
+    ) -> Result<Vec<C::Scalar>, Error> {
+        let blinding_factors = pk.vk.cs.blinding_factors();
+        let challenges: Vec<C::Scalar> = vec![*beta, *gamma, *delta];
+        let mut shuffle_product = vec![C::Scalar::one(); params.n as usize];
+
+        // #[cfg(not(feature = "cuda"))]
+        {
+            let expression_group_product =
+                |expression_group: &[Polynomial<C::Scalar, LagrangeCoeff>],
+                 products: &mut [C::Scalar]| {
+                    for (expression, challenge) in expression_group.iter().zip(challenges.iter()) {
```

Comment:
This challenge design limits the group size up to 3. I believe 3 is the maximal value for zkWasm, but in gereral it is not. You can just follow the design of permutation. To seperate the chellange for different shuffle in the group, we just need to multiply a constant `DELTA` on the challenge.
Consider from the attacker side. If we have an unknown challenge value `beta`, and three shuffle pairs `[a0, a1], [b0, b1], [c0, c1]` in a group. We require
`(a0 + beta) * (b0 + beta * DELTA) * (c0 + beta * DELTA * DELTA) == (a1 + beta) * (b1 + beta * DELTA) * (c1 + beta * DELTA * DELTA)`.
Can we find `[a2, b2, c2] != [a1, b1, c1]` that
`(a0 + beta) * (b0 + beta * DELTA) * (c0 + beta * DELTA * DELTA) == (a2 + beta) * (b2 + beta * DELTA) * (c2 + beta * DELTA * DELTA)`?

Reply:
Subtly，currently different challenges  just refer to lookup's grant product (A+beta)*(S+gamma), in my opinion, what the totally random beta and gamma will make A and S product totally independently.
Considering permutation's  (A+delta^i*beta+gamma)*(B+delta^j*beta+gamma), the beta just make the A/B independently with delta^i/delta^j, and the gamma just to challenge the grant product to numerator and denominator ， (A+delta^i*beta+gamma)*(B+delta^j*beta+gamma) and  (A‘+s(i)*beta+gamma)*(B+s(j)*beta+gamma), here A and B can be considered as a whole， so one gamma is enough。
 
As different shuffle group is independent，they can not be considered as a whole， so one challenge  as permutation‘s gamma is not enough。Refer to lookup's different challenge may be the safest.

Reply:
Yes, I think your solution is correct. I am trying to prove:
For group size = 3, we have
```
(a + x)(b + x^2)(c + x^3) = x^6 + ax^5 + bx^4 + (ab + c)x^3 + ...,
So a, b can be constrained by coeff of ax^5 + bx^4. Since (a, b) is guaranteed, c can be constrained by (ab + c)x^3
```
Suppose n-size group can be constrained.
Let's see group size = n + 1
```
A * (vn + x^n) has same n leading coeff of A.
So v0..vn-1 can still be constrained by the leading n terms, and thus vn can be constrained by other the terms.
```